### DISCUSSION TITLE
U8g2 MUI organisation as part of a custom UIManager class

### DISCUSSION THREAD

**Discussion by maarten1C96** [olikraus/u8g2#2252] [Q&A]
Using the U8g2 library with the MUI addon has been a great pleasure so far. Initially I wrote all my code in one single main.cpp file which was working fine. But when my code started growing beyond pleasantly scrollable I chose to organise everything into different classes, rewriting my code from the ground up. I am now trying to construct a class called UIManager which is supposed to be responsible for handling all UI input (rotary encoder, buttons, etc.) and controlling UI output (SSD1306 OLED, different LEDs, alarm buzzer, etc.). Currently this class is able to display buttons, labels, and other more static data on my OLED, but since I started using "MUIF_RO" within the MUIF table,  I've encountered an error that says:

_a value of type "uint8_t (UIManager::*)(mui_t *ui, uint8_t msg)" cannot be used to initialize an entity of type "muif_cb"_

Using the same function in my initial, all-included main.cpp was working fine. And also when commenting out the specific line in my current UIManager.cpp, responsible for linking the function with a MUI item in my 'muif_list' (`MUIF_RO("SH", show_my_data)`), it works fine again as well. Since MUI makes use of preprocessor directives, is it even possible to organise custom MUI function callbacks from within a separate UIManager class?

(I am relatively new to C++ programming)

**Comment by olikraus** [comment]
"muif_cb" is defined as 
``` C++
typedef uint8_t (*muif_cb)(mui_t *ui, uint8_t msg);
```
You function must exactly look like this, especially it must not be a member function of a class (like in your example).
You can however use a friend function of your class if it has the same arguments.
Moreover I assume you need to create a global UIManager object, which is then accessed by the friend function, which in turn is used as MUI callback.

**Comment by PhilippMolitor** [comment]
I'm having the exact same issue. I noticed that first when MUI rendered just a blank screen when initializing the Fds and Fpl inside of a class method. It would be great to have an example of how to use MUI with class methods, or a workaround with friend functions.

Also for my code, I used `std::function<T>` instead of a `typedef returnType (*cb)();` which allows to use lambdas and all sorts of other stuff for callbacks. Maybe this could be an option, if it is portable enough?

**Comment by olikraus** [comment]
The core code of u8g2 is written in C only. I continued doing so with MUI. 
The C++/Arduino code is just a wrapper around the C code. 

If you want me to create a C++ example, then maybe consider the creation of an issue... but my time is a little bit limited...

**Reply by PhilippMolitor** [reply]
Thanks for your efforts, I appreciate it. I wrote a quick summary of my thoughts in #2258

**Reply by olikraus** [reply]
ok, let's continue the discussion in the issue

**Reply by maarten1C96** [reply]
@olikraus Yes, thanks you for your efforts as well (and for your prompt replies)! I am again trying to carefully rewrite my own code and wrapping my head around your initial answer (and the logic of my overall code that I'm trying to achieve in general). Once I believe to have found some working solution myself I'll post this, but I am not that fast!

**Comment by maarten1C96** [comment]
So far, the following structure of code seems to work for me: 

- Passing _references_ to the relevant UI objects (`&u8g2`, `&mui`, `&r`) via the constructor of my new UIManager class.
- Passing `muif_list` and `fds_data` (as values) into the `begin()` method of this _UIManager class_.
- Calculating the number of entries in the MUIF list in `setup()` and passing the _resulting value_ to the `begin()` method (instead of calculating that value _within_ the UIManager class, which I actually would prefer but wasn't able to make functional so far).

I included a screenshot of the structure in my _main.cpp_ (at the end of this post), meanwhile the following is how my constructor currently handles references in _UIManager.cpp_:

```
UIManager::UIManager(U8G2_SSD1306_128X64_NONAME_F_HW_I2C* u8g2, MUIU8G2* mui, Rotary* r)
    : _u8g2(u8g2), _mui(mui), _r(r)
{
    // constructor yet only for passing references to 
}
```


...and how my `begin()` method in _UIManager.cpp_ handles `muif_list` and `fds_data` with the different begin()-methods included in the u8g2 library:

```
void UIManager::begin(muif_t* muif_list, fds_t* fds_data, size_t muif_list_size) {
    _fds_data = fds_data;
    _muif_list = muif_list;
    _muif_list_size = muif_list_size;
    _u8g2->begin();  
    _mui->begin(*_u8g2, _fds_data, _muif_list, _muif_list_size);
    _mui->gotoForm(1, 0); // form_id, initial_cursor_position
}
```


This is my result so far, although I didn't yet fully achieve my intentions of making _main.cpp_ maximally clean yet though. For moving the header files for the MUI callback functions and `muif_list` and `fds_data` in _main.cpp_ I am currently thinking about the following:

-  Including those MUI callback functions files within _UIManager.h_, and meanwhile instead of the conventional position of `#include "UIManager.h"` at the top of the page, moving this down to just before the initialisation of the UIManager object (`UIManager ui_manager(&u8g2, &mui, &r)`).

- Puzzling with declaring an array of multiple pointers, positioned just before `muif_list` (for reasons of the preprocessor directives), whose pointers could then later be updated with new addresses in the UIManager class (double forwarded locations of functions)? Even it this will work, it will probably also result in many unused pointers (which could be avoided by manual counting though, but hassle).

- I still need to look deeper into the 'friend function' mentioned by @olikraus. Although very interesting I didn't yet figured out how to optimally utilise these privileged interactions within my context.

- Or accept the current header files as they are, which do not seem to be so negatively affecting the readability yet, although in my specific situation the code will eventually grow much further (when having to handle different sensors, motors, more classes, etc.), resulting in an unpleasantly overcrowded _main.cpp_ file.

I would be happy to receive critical remarks and general thoughts (and as mentioned before please know I am not an expert in this)!

![Screenshot 2023-09-03 at 14 44 24](https://github.com/olikraus/u8g2/assets/12756980/c6e303fd-b90a-45ca-9542-525ab310e320)

**Reply by maarten1C96** [reply]
For those interested, this is my working code so far (including all header and source files): [granulaatblazer_efef70b9.zip](https://github.com/olikraus/u8g2/files/12506579/granulaatblazer_efef70b9.zip)

**Reply by PhilippMolitor** [reply]
In #2258 you can see my example which stores the MUI variables inside member variables of my UI class. I also recommended a way to modify the callback type of the MUI library to work with member functions, but there still is some figuring out to be done.

**Reply by typ127** [reply]
@maarten1C96 Thanks a lot for the zip! This saved my day. I tried to solve this for at least one week now. Finaly found your post.

**Reply by maarten1C96** [reply]
@typ127 Happy to hear :-)

**Comment by rushairer** [comment]
Mark
